pinot-core/src/main/java/org/apache/pinot/core/data/table/ConcurrentIndexedTable.java

```diff
@@ -101,31 +108,51 @@ public int size() {
 
   @Override
   public Iterator<Record> iterator() {
+    if (_sort && CollectionUtils.isNotEmpty(_orderBy)) {
+      List<Record> sortedList = new ArrayList<>(_lookupMap.values());
+      sortedList.sort(_orderByComparator);
+      return sortedList.iterator();
+    }
     return _lookupMap.values().iterator();
   }
 
   private void resize(int trimToSize) {
 
     if (_lookupMap.size() > trimToSize) {
 
-      // make min heap of elements to evict
-      int heapSize = _lookupMap.size() - trimToSize;
-      PriorityQueue<Record> minHeap = new PriorityQueue<>(heapSize, _minHeapComparator);
-
-      for (Record record : _lookupMap.values()) {
-        if (minHeap.size() < heapSize) {
-          minHeap.offer(record);
-        } else {
-          Record peek = minHeap.peek();
-          if (minHeap.comparator().compare(record, peek) < 0) {
-            minHeap.poll();
+      if (CollectionUtils.isNotEmpty(_orderBy)) {
+        // drop bottom
+
+        // make min heap of elements to evict
+        int heapSize = _lookupMap.size() - trimToSize;
+        PriorityQueue<Record> minHeap = new PriorityQueue<>(heapSize, _minHeapComparator);
+
+        for (Record record : _lookupMap.values()) {
+          if (minHeap.size() < heapSize) {
             minHeap.offer(record);
+          } else {
+            Record peek = minHeap.peek();
+            if (minHeap.comparator().compare(record, peek) < 0) {
+              minHeap.poll();
+              minHeap.offer(record);
+            }
           }
         }
-      }
 
-      for (Record evictRecord : minHeap) {
-        _lookupMap.remove(evictRecord.getKey());
+        for (Record evictRecord : minHeap) {
+          _lookupMap.remove(evictRecord.getKey());
+        }
+      } else {
+        // drop randomly
+
+        int numRecordsToDrop = _lookupMap.size() - trimToSize;
```

Comment:
its better to not add new keys if there is no orderby rt

Reply:
True. Would it be better to make the Operator keep track and stop adding more? Or let the operator keep adding everything, but make the table intelligent to ignore them?
Pro of the first approach is we don't add extra records. However, the operators have to do special handling depending on order by/no order by.
Pro of the second approach is operators need not know anything special. All intelligence is at the table. However, we will read some extra rows, create Record object, Key object, spend time converting raw values, only to ignore the values.

Seems like 1st would be more efficient.

Reply:
not sure if we can simply ignore new rows once the limit is reached. We can only ignore new keys, we still need to update values for existing keys right?

Reply:
Ahh yes.. In that case, we either need the table to support `get(key)`, or we need to make the upsert check `if (isNewKey && isCapacityFull && !isOrderBy)`

Reply:
Handled this. Please take a look. @mayankshriv FYI